Snowflake.Data/Core/SFDataConverter.cs

```diff
@@ -20,74 +20,100 @@ public enum SFDataType
+            byte[] b1 = Encoding.UTF8.GetBytes(srcVal);
+            byte[] b2 = new byte[b1.Length + 100];
+            Array.Copy(b1, 0, b2, 100, b1.Length);
+            var v = new UTF8Buffer(b2, 100, b1.Length);
+            return ConvertToCSharpVal(v, srcType, destType);
         }
-        
-        internal object ConvertToCSharpVal(string srcVal, SFDataType srcType, Type destType)
-        {
-            //Logger.DebugFmt("src value: {0}, srcType: {1}, destType: {2}", srcVal, srcType, destType);
 
+        internal static object ConvertToCSharpVal(UTF8Buffer srcVal, SFDataType srcType, Type destType)
+        {
             if (srcVal == null)
                 return DBNull.Value;
 
-            if (destType == typeof(short) 
-                || destType == typeof(int)
-                || destType == typeof(long)
-                || destType == typeof(Guid)
-                || destType == typeof(double)
-                || destType == typeof(float)
-                || destType == typeof(decimal))
+            try
             {
-                if (!typeConverters.ContainsKey(destType))
+                // The most common conversions are checked first for maximum performance
+                if (destType == typeof(Int64)) 
                 {
-                    typeConverters.Add(destType, TypeDescriptor.GetConverter(destType));
+                    return FastParser.FastParseInt64(srcVal.Buffer, srcVal.offset, srcVal.length);
+                }
+                else if (destType == typeof(Int32))
+                {
+                    return FastParser.FastParseInt32(srcVal.Buffer, srcVal.offset, srcVal.length);
+                }
+                else if (destType == typeof(decimal))
+                {
+                    return FastParser.FastParseDecimal(srcVal.Buffer, srcVal.offset, srcVal.length);
+                }
+                else if (destType == typeof(string))
+                {
+                    return srcVal.ToString();
+                }
+                else if (destType == typeof(DateTime))
+                {
+                    return ConvertToDateTime(srcVal, srcType);
+                }
+                else if (destType == typeof(DateTimeOffset))
+                {
+                    return ConvertToDateTimeOffset(srcVal, srcType);
+                }
+                else if (destType == typeof(Boolean))
+                {
+                    return srcVal.Buffer[0] == '1' || srcVal.Buffer[0] == 't' || srcVal.Buffer[0] == 'T';
+                }
+                else if (destType == typeof(byte[]))
+                {
+                    return srcType == SFDataType.BINARY ?
+                        HexToBytes(srcVal.ToString()) : srcVal.GetBytes();
+                }
+                else if (destType == typeof(Int16))
+                {
+                    return (Int16)FastParser.FastParseInt32(srcVal.Buffer, srcVal.offset, srcVal.length);

```

Comment:
This will not catch overflows and return erroneous data to the user instead of an error.

Reply:
Good catch! I thought a narrowing cast generated overflow, but you are correct.
Fixed